controller/array_action/ds8k_rest_client.py

```diff
@@ -0,0 +1,297 @@
+from pyds8k.client.ds8k.v1.client import Client
+from pyds8k.exceptions import NotFound
+from controller.common.csi_logger import get_stdout_logger
+
+logger = get_stdout_logger()
+
+
+def dictionarize(func):
+    def inner(self, *args, **kwargs):
+        res_obj = func(self, *args, **kwargs)
+        if not isinstance(res_obj, list):
+            res_obj = [res_obj, ]
+        coverted = []
+        for res in res_obj:
+            try:
+                coverted.append(res.representation)
+            except AttributeError:
+                coverted.append(res)
+        return coverted
+    return inner
+
+
+def _int_lunid_to_hex(lunid):
+    return '{0:0{1}x}'.format(int(lunid), 2)
+
+
+def int_to_scsilun(lun):
+    # There are two style lun number, one's decimal value is <256 and the other
+    # is full as 16 hex digit. According to T10 SAN, when decimal value value
+    # is more then 256 and it is converted to the full 16 hex digit, it should
+    # be swapped and hex.
+    # For example, zlinux lun number stored in SC DB is 1075331113. It should
+    # be hex to '40184029' and swapped to '40294018'. When the lun number is 12
+    # and it is <256, it should be hex to '0c' directly.
+    pretreated_lun = int(lun)
+    if pretreated_lun < 256:
+        return _int_lunid_to_hex(pretreated_lun)
+    else:
+        return '{0:x}'.format(
+            (pretreated_lun >> 16 & 0xFFFF) | (pretreated_lun & 0xFFFF) << 16
+        )
+
+
+def scsilun_to_int(lun):
+    # There are two style lun number, one's decimal value is <256 and the other
+    # is full as 16 hex digit. According to T10 SAN, the full 16 hex digit
+    # should be swapped and converted into decimal.
+    # For example, SC got zlinux lun number from DS8K API, '40294018'. And it
+    # should be swapped to '40184029' and converted into decimal, 1075331113.
+    # When the lun number is '0c' and its decimal value is <256, it should be
+    # converted directly into decimal, 12.
+    pretreated_scsilun = int(lun, 16)
+    if pretreated_scsilun < 256:
+        return pretreated_scsilun
+    else:
+        return (pretreated_scsilun >> 16 & 0xFFFF) | \
+               (pretreated_scsilun & 0xFFFF) << 16
+
+
+class RESTClient(object):
```

Comment:
Isn't it the class where you copied [all this code](https://github.com/IBM/pyds8k/blob/75c6af1fc49f62d4969df8ea66913f838ee0e5d0/pyds8k/client/ds8k/v1/sc_client.py#L26) from?

Reply:
How do you know that it's just an example? It looks like an official public code.
Either way, most of the code is identical, so there is no reason not to use it!

Reply:
Since you wrote all relevant classes, and your design doesn't allow `RESTClient` to inherit from `SCClient`, BUT still most of the logic of `SCClient` is useful for `RESTClient`, then a fix, for example, is a middle class containing this logic (that inherits from `Client`), that both `SCClient` and `RESTClient` will inherit from.
There is no need to duplicate class code, that's what inheritance is for.

Reply:
In fact, the Client itself is powerful and flexible enough (you can see most of the methods just call the Client and return). I don't want a client in client structure as a middle class.
But if you think that some methods are complex and worth defining in Client, please point them out, I will consider to add them.
One example I can see is the create_volume method.

Reply:
@27149chen I don't mean client in client structure (e.g. `RESTClient` will contain `Client` member), I mean inherit (e.g. `class RESTClient(Client):`). there is no need for methods that only pass the same arguments to another method.

also no need to implement `normalize_response`, because you can import `dictionarize`.
same for all the consts...

Reply:
Sorry, I don't want to use inherit here, because I don't want the RESTClient consumers to access the Client methods directly.
Please forget the SCClient in pyds8k.

Reply:
1. Why don't you want them to access the `Client` methods "directly"? how is it different than simply passing everything to the `self.client` method? if some method is missing some logic, it can always be overridden.
2. What about the following format?:
```
class RESTClient(Client):
.
.
.
def get_system(self):
        return super().get_system()
.
.
.
```

Reply:
1. Because I want to isolate the underlying implementation and expose only limited interfaces.
2. no, it is meaningless.

Reply:
there are many functions which are not visible in RESTClient, like get_marrays, get_io_enclosures, ...
The RESTClient should only use the functions defined in this file.

Reply:
OK